crates/iceberg/src/io.rs

```diff
@@ -278,7 +288,7 @@ impl OutputFile {
     }
 
     /// Creates output file for writing.
-    pub async fn writer(&self) -> Result<impl FileWrite> {
+    pub async fn writer(&self) -> Result<Writer> {
         Ok(self.op.writer(&self.path[self.relative_path_pos..]).await?)
```

Comment:
It's better to add a `buffer` for this writer to avoid `EntiryTooSmall` error from s3.

---

There is another thing we can do is adding `concurrent` here so we can upload parts concurrently. But we need to make a design on how users control the concurrency. Maybe we can discuss them later.

Reply:
> There is another thing we can do is adding concurrent here so we can upload parts concurrently. But we need to make a design on how users control the concurrency. Maybe we can discuss them later.

Yes, we can open a discussion about it. IMO, the concurrent writer is used in the case that writes multiple files concurrently. So concurrent writer may should be a higher-level writer.

Reply:
> It's better to add a buffer (for example, 8MiB) for this writer to avoid EntiryTooSmall error from s3.

> And in the future, we can add a config for this to allow users to control them.

Sounds reasonable for me. So this means that the writer in opendal will not provide the buffer cache, right?

Reply:
> Yes, we can open a discussion about it. IMO, the concurrent writer is used in the case that writes multiple files concurrently. So concurrent writer may should be a higher-level writer.

Yes, there are two levels of concurrency here. We can write different files concurrently, and we can also write different parts of the same file concurrently. The second case is particularly relevant when writing a large parquet file, such as 10GiB or even 1TiB.

However, we don't need to delve into such details at the current stage. I just want to note that there are potential optimizations possible. Should we leave a todo here?

Reply:
> Sounds reasonable for me. So this means that the writer in opendal will not provide the buffer cache, right?

opendal doesn't add buffer for users by default. Users will need to use `op.writer_with(path).buffer(8 * 1024 * 1024)` to enable it.

Reply:
I find that the `AsyncArrowWriter` already provides the [write with a buffer](https://docs.rs/parquet/latest/parquet/arrow/async_writer/struct.AsyncArrowWriter.html#method.write ). In this case, maybe we can use this buffer.

And we also provide the interface  `pub async fn writer(&self, buffer_size: usize) -> Result<Box<dyn FileWrite>>` in OutFile, for the file format which doesn't provide the cache, we can create a inner writer with buffer.

Reply:
> I find that the `AsyncArrowWriter` already provides the [write with a buffer](https://docs.rs/parquet/latest/parquet/arrow/async_writer/struct.AsyncArrowWriter.html#method.write). In this case, maybe we can use this buffer.

Makes sense,